### PR TITLE
Add a robustier `instanceof` alternative to allow cross-package compatibility.

### PR DESCRIPTION
**Pull request by Shaptic**
**Note:** Lots of people tagged for review mostly as an FYI as there may or may not be far-reaching implications to this change.

### What
Introduce a better check for cross-compatible `XdrCompositeType` subtypes (`Union`, `Struct`, `Enum`, and `Array`) based on [a Discord discussion](https://discord.com/channels/897514728459468821/1082043429053276190/1196889142072717462) from @earrietadev which will check the prototype chain for the constructors and properties we need:

```typescript
export function isSerializableIsh(
    value: any, 
    subtype: InstanceType<typeof XdrCompositeType>
): value is XdrCompositeType;
```

(This isn't a proper TypeScript definition because js-xdr (a) doesn't use TS and (b) doesn't use interfaces, but imagine `XdrCompositeType` as being an interface and it makes more sense.)

We also make the error messages way better.

#### Details
Because the classes are generated, their prototype constructor name doesn't match the actual class name more often than not. However, they do still inject the "true" name into their prototype. The three types we're trying to match on all do this differently:
 * `struct`s will use `structName`, see [struct.js](https://github.com/stellar/js-xdr/blob/master/src/struct.js#L45)
 * `union`s will use `unionName`, see [union.js](https://github.com/stellar/js-xdr/blob/master/src/union.js#L101)
 * `enum`s will use `enumName`, see [enum.js](https://github.com/stellar/js-xdr/blob/master/src/enum.js#L71)

It's not perfect, but it's better than the `instanceof` check we have now.

So we will check these fields as well as the existence of a `read()` and `write()` method to make a best-effort guess that the given value matches the kind we expect.

### Why
Major versions of a package should probably be cross-compatible in the same dependency tree. For example, `js-xdr@3.1.1` should function interchangeably alongside `js-xdr@3.0.0` if the additive features are not relevant.

That is not the case today due to `instanceof` checks that become package-specific, while `isSerializableIsh` should rectify this.

### Known Limitations
There are _only_ performance implications for people who are introducing this mixing behavior (besides the new overhead of a function call which I would consider trivial), so there are no performance regression risks.

### CODE REVIEW

**Comment by BlaineHeffron** [comment]
The new check also fails to resolve https://github.com/stellar/soroban-cli/issues/1285

File: src/xdr-type.js [hunk 1]
```diff
@@ -170,6 +170,48 @@ function decodeInput(input, format) {
...
+ * meaning serialization, but also conditioned on a particular subclass of
+ * "XdrCompositionType" (e.g. {@link Union} which extends XdrCompositeType ->
+ * XdrType).
+ *
+ * This makes the package resilient to downstream systems that may be combining
+ * many versions of a package across its stack that are technically compatible
+ * but fail `instanceof` checks due to cross-pollination.
+ */
+export function isSerializableIsh(value, subtype) {
+  const hasConstructor = (instance, name) => {
```
Comment (willemneal) [inline src/xdr-type.js:189]: This doesn't capture any context of the containing function. And seems useful on its own.

**Review by willemneal — APPROVED** [review]
One small nit, but I don't think it should prevent merging. Always fun digging into the guts of JS.

File: src/xdr-type.js (hunk 1 above)
Comment (Shaptic) [reply src/xdr-type.js]: Good point!

File: src/enum.js [hunk 2]
```diff
@@ -26,8 +26,13 @@ export class Enum extends XdrPrimitiveType {
    * @inheritDoc
    */
   static write(value, writer) {
-    if (!(value instanceof this))
-      throw new XdrWriterError(`unknown ${value} is not a ${this.enumName}`);
+    if (!this.isValid(value)) {
```
Comment (willemneal) [inline src/enum.js:29]: I'm assuming that `this` is the Class, but personally I find it less confusing if you use the Class directly to remove any ambiguities. Plus the caller could have used `bind` to change this method's `this`. Not that we need to worry, but it's possible.
```suggestion
    if (!Enum.isValid(value)) {
```

**Review by willemneal — APPROVED** [review]
I've always had issues with `this` in JS and in `this` case being `this` as the Class or (constructor?) is a bit odd to me. But if this is the standard JS way, pay my reservations to heed.

**Comment by piyalbasu** [comment]
This looks solid to me. I'll echo @willemneal 's thought about using the Class rather than `this`. Probably not super critical, but you may as well cover all your bases

**Review by piyalbasu — APPROVED** [review]

File: src/enum.js (hunk 2 above)
Comment (Shaptic) [reply src/enum.js:29]: Unfortunately, we have to use `this` here for an oddly-specific reason: we need to make sure it deals with child classes and any modifications that are made after creation, e.g. like [here](https://github.com/stellar/js-xdr/blob/master/src/union.js#L99-L102). Just passing the class name doesn't cover this case.

Comment (willemneal) [reply src/enum.js:29]: But shouldn't the classed be sealed? Why do we need to consider the case of subclasses?

Comment (Shaptic) [reply src/enum.js:29]: You can see in that linked line that a `Union` itself is never actually created, it's always a locally-generated subclass of `ChildUnion` that has a bunch of properties attached to it (most importantly `unionName`) that gets returned. So we'd need to be able to `[the specific ChildUnion class].isValid` if we wanted an accurate evaluation of "this"ness.

**Comment by Shaptic** [comment]
I've tried testing this with downstream systems and landed upon the wonderfully helpful "[object Object]'s struct name is SorobanAuthorizationEntry not SorobanAuthorizationEntry: <a JSON dump of the object>" so clearly the extra check is still failing in key moments. I'm still in the process of debugging it and will merge this once I can pull that off.

**Comment by Shaptic** [comment]
Yooooo I ran this through a project that uses `vite` all the way down through the SDK and it works!! :partying_face: :partying_face:

**Pull request merged**